feat: Add the last merged table to the public inputs of the hiding circuit

federicobarbacovi opened the pull request:
We make the merged table received by the Merge verifier in the hiding circuit a public input to the hiding circuit. This is needed because the Merge verifier will soon receive `t_commitments`, `T_prev_commitments` as inputs rather than reading them from the proof.

**EDIT:**

To complete the work on the consistency checks, and to ensure the soundness of the Goblin verification, the merged table received by the Merge verifier in the last step of a Goblin accumulation must be set to be a public input of the circuit that performs the verification, so that the verifier can extract that public input and use it as the commitment to the previous table in the Merge verification.

For example, in ClientIVC the last Merge verification before the final Goblin verification happens in the HidingKernel, so we need to add the merged table commitments received by the Merge verifier inside the HidingKernel to be public inputs of the HidingKernel.

After this PR,  `MegaVerifier = UltraVerifier<MegaFlavor>` always expects the inputs to be `PairingInputs` + commitments to ECC op tables. These inputs are produced by the class `HidingKernelIO` (even though in the future we might consider changing this name)

The PR required changes to various tests to accommodate the new structure of the public inputs.

Commit:
[empty] Start merge-train. Choo choo.

ledwards2225 committed:
chore!: Remove circuit_size from all VKs (#15747)

- Removes `circuit_size` from all VKs (only store `log_circuit_size`)
- Updates `compute_public_input_delta()` to correctly take witness
values and compute `dyadic_size` via `pow()`

Addresses most of
https://github.com/AztecProtocol/barretenberg/issues/1283 but one
instance remains

federicobarbacovi committed:
chore: Introduce data structures to hold inputs/outputs of the Merge verification (#15735)

Introduce the classes `SubtableWitnessCommitments` which stores
`t_commitments` (and in the future `T_prev_commitments`) and
`WitnessCommitments`, which stores `t_commitments`, `T_commitments` (and
in the future `T_prev_commitments`). The role of these classes is to
facilitate the introduction of consistency checks between two
consecutive merges, see
[#1351](https://github.com/AztecProtocol/barretenberg/issues/1351).

- The value of `t_commitments` is populated by extracting the
commitments from the VK of the PG verifier.
- (Future work) The value of `T_prev_commitments` is populated by taking
the previous value of `T_commitments`
- The value of `T_commitments` is populated by the Merge verifier

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: notnotraju <raju@aztec-labs.com>
Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>
Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Santiago Palladino <spalladino@users.noreply.github.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
Co-authored-by: ludamad <domuradical@gmail.com>
Co-authored-by: maramihali <mara@aztecprotocol.com>

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

johnathan79717 committed:
chore: replace asserts with runtime errors. (#15671)

1. Asserts now throw runtime errors in non-debug builds.
1. `BB_ASSERT_*` now throw_or_abort in non-debug builds.
2. Replace `ASSERT` in tests with `ASSERT_TRUE` or other gtest
functions.
3. Replace `EXPECT_DEATH` and `ASSERT_DEATH` in tests with
`EXPECT_THROW_OR_ABORT` or `ASSERT_THROW_OR_ABORT`.
4. Remove `#ifdef NDEBUG` around the above in tests since they work in
non-debug builds now.
5. Asserts in `constexpr`s uses info instead of sstream.

Fixes https://github.com/AztecProtocol/barretenberg/issues/1460

Commit:
chore: kernels start with eq and reset (#15734)

As a measure of precaution in case of misconfigured ecc ops in an app
circuit that might attempt, for example, to link ecc ops between apps
and kernels, the subtable of ecc ops in each kernel circuit will start
with an eq and reset op. This cannot happen when the API for creating
ecc ops is used properly as the logic in `goblin_element` ensures an eq
and reset op is present at the end of each `batch_mul`, function also
used for `operator+` and `operator-`. Moreover, the
`add_gates_to_ensure_nonzero_polynomials` makes sure that an eq and
reset exists if no other ecc ops are encountered in the circuit.

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: notnotraju <raju@aztec-labs.com>
Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>
Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Santiago Palladino <spalladino@users.noreply.github.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
Co-authored-by: ludamad <domuradical@gmail.com>
Co-authored-by: Sarkoxed <75146596+Sarkoxed@users.noreply.github.com>

federicobarbacovi committed:
Introduce hiding kernel io class

federicobarbacovi committed:
Update hiding kernel io to contain array of commitments

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Update civc

federicobarbacovi committed:
Update ClientIVC recursive verifier

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Fix test

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Squashed commit of the following:

commit 2a5f380499bad7e2b8e2ab633b2aced6f7fe9f89
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Fri Jul 18 11:25:25 2025 +0000

    Fixes

commit 62c726005a5daa6c54069b60ad208e69bc6679e8
Merge: f96e42c7df ae67a84745
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Fri Jul 18 11:20:24 2025 +0000

    Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/native_io_mechanism

commit f96e42c7df4bb04b6a37e28afefec60dc637e116
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Fri Jul 18 11:17:55 2025 +0000

    Introduce native IO mechanism and update ultra verifier interface

commit 9c85f50c86a46973398985848291e0566ba93ac2
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Fri Jul 18 09:43:37 2025 +0000

    Changes to function signatures

commit 1885708d570fb27877815a3c0546f3515ef65e28
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Thu Jul 17 18:00:17 2025 +0200

    feat: Implement native reconstruction from public inputs for `AffineElement`, `bb::fq`, and `bb::fr` (#15774)

    We make public input reconstruction in native land more idiomatic by introducing `reconstruct_from_public` method for `fq`, `fr`, and `affine_element`.

    ---------

    Co-authored-by: AztecBot <tech@aztecprotocol.com>
    Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
    Co-authored-by: ludamad <adam.domurad@gmail.com>
    Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
    Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
    Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>
    Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
    Co-authored-by: notnotraju <raju@aztec-labs.com>
    Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
    Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>
    Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
    Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
    Co-authored-by: Santiago Palladino <spalladino@users.noreply.github.com>
    Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
    Co-authored-by: ludamad <domuradical@gmail.com>
    Co-authored-by: maramihali <mara@aztecprotocol.com>
    Co-authored-by: Sarkoxed <75146596+Sarkoxed@users.noreply.github.com>

commit d2184811d97a1478d82a115c3be1e34d888b4c14
Merge: 5380e5505c afc8002ee4
Author: AztecBot <tech@aztecprotocol.com>
Date:   Thu Jul 17 15:14:28 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 5380e5505c9f88a192a8499eeaad54d1dca7eb5a
Merge: 9ac2ef892e 4023da43e2
Author: AztecBot <tech@aztecprotocol.com>
Date:   Thu Jul 17 14:53:32 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 9ac2ef892e392c2ebbe111f39d9713dd24a7f34e
Merge: 8b7150994c 710c7919a1
Author: AztecBot <tech@aztecprotocol.com>
Date:   Thu Jul 17 11:48:55 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 8b7150994c9ecad6961156cc1fe5370a9c996cb5
Merge: 7ea4056351 4aeb094995
Author: AztecBot <tech@aztecprotocol.com>
Date:   Thu Jul 17 09:01:52 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 7ea4056351cb3a4cd03d21d56de22776b83cc8c3
Merge: ad3dea50ce 980391ab71
Author: AztecBot <tech@aztecprotocol.com>
Date:   Thu Jul 17 06:19:25 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit ad3dea50cea0081ab64730f2f2774f1c5eeda547
Merge: 8456facc2d 0ab34eefd0
Author: AztecBot <tech@aztecprotocol.com>
Date:   Wed Jul 16 22:51:35 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 8456facc2d176abe1aec2ee6b8938e0de15885ef
Merge: de7c071eb2 445db95628
Author: AztecBot <tech@aztecprotocol.com>
Date:   Wed Jul 16 22:26:26 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit de7c071eb2280be82241da5fb0465cdd18867489
Merge: 0968cf0f20 9bbff1b82f
Author: AztecBot <tech@aztecprotocol.com>
Date:   Wed Jul 16 22:04:07 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 0968cf0f20a03699017173220cffc5dafced2d47
Merge: 921cd0343b 69bcc536a9
Author: AztecBot <tech@aztecprotocol.com>
Date:   Wed Jul 16 21:57:40 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 921cd0343bb97e1103f79bb09f543f0afcba19f5
Merge: 29c5a1fbf6 03138c41af
Author: AztecBot <tech@aztecprotocol.com>
Date:   Wed Jul 16 18:45:54 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 29c5a1fbf60c1e27f3a17148cf17eea45dea7c5e
Merge: 6cd32ca589 6205381328
Author: AztecBot <tech@aztecprotocol.com>
Date:   Wed Jul 16 17:59:18 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 6cd32ca589cd1f2fe310db3e2dc247cda96ba57f
Merge: 6893736250 1891f8d05e
Author: AztecBot <tech@aztecprotocol.com>
Date:   Wed Jul 16 16:55:29 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 6893736250f6d80ba37598b4bb02d77055829961
Merge: 9f27abbf32 5988a3553b
Author: AztecBot <tech@aztecprotocol.com>
Date:   Wed Jul 16 16:44:00 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit 9f27abbf320f3e0ea1a3d80ffe01d954c9650154
Author: AztecBot <tech@aztecprotocol.com>
Date:   Wed Jul 16 16:21:05 2025 +0000

    [empty] Start merge-train. Choo choo.

federicobarbacovi committed:
Remove flavor

federicobarbacovi committed:
Fix

federicobarbacovi committed:
Delete bb-civc-inputs-tar.gz

federicobarbacovi commented on barretenberg/cpp/bb-civc-inputs.tar.gz:
Removed as I imagine it was added by mistake

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
chore: Introduce Native IO mechanism (#15820)

We introduce native counterparts for our IO mechanism so that the
reconstruction of public inputs in from native field elements is similar
to that in-circuit

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: notnotraju <raju@aztec-labs.com>
Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>
Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Santiago Palladino <spalladino@users.noreply.github.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
Co-authored-by: ludamad <domuradical@gmail.com>
Co-authored-by: maramihali <mara@aztecprotocol.com>
Co-authored-by: Sarkoxed <75146596+Sarkoxed@users.noreply.github.com>

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/add_merge_table_to_hiding_circuit_pub_inputs

federicobarbacovi committed:
Move native logic out of stdlib; update tests

federicobarbacovi committed:
Update tests

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/stdlib/primitives/biggroup/biggroup.hpp:
```diff
@@ -21,8 +21,9 @@
-template <class Builder, class Fq, class Fr, class NativeGroup> class element {
+template <class Builder_, class Fq, class Fr, class NativeGroup> class element {
   public:
+    using Builder = Builder_;
```
Needed so that we can use `PublicComponentKey` on `element`

federicobarbacovi committed:
Fixes

Commit:
[empty] Start merge-train. Choo choo.

johnathan79717 committed:
chore: flag to update inputs when precomputed vk fails (#15828)

This is the first part to fix
https://github.com/AztecProtocol/barretenberg/issues/1464. Once we have
this, we can add an option to
`test_civc_standalone_vks_havent_changed.sh` that generates inputs with
this option.

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/add_merge_table_to_hiding_circuit_pub_inputs

federicobarbacovi committed:
Remove inputs

ledwards2225 commented on barretenberg/cpp/src/barretenberg/stdlib/client_ivc_verifier/client_ivc_recursive_verifier.cpp:
```diff
@@ -25,6 +25,10 @@ ClientIVCRecursiveVerifier::Output ClientIVCRecursiveVerifier::verify(const Stdl
     MegaVerifier::Output mega_output = verifier.verify_proof(proof.mega_proof);
 
+    // Extract public inputs
+    bb::stdlib::recursion::honk::HidingKernelIO<Builder> hiding_output;
```
This actually seems a little funny. Am I right that the pairing points are being reconstructed twice, once within the Honk recursive verifier as part of `DefaultIO` and again here? I guess the idea is they'll get ignored on this second reconstruction but it would be nice to unify. Could we simply perform this `HidingKernelIO` reconstruction in the ultra honk recursive verifier conditionally based on `IsMegaFlavor`? Might wreak havoc in certain tests..

federicobarbacovi replied:
Right, I had not noticed that. I create this [issue](https://github.com/AztecProtocol/barretenberg/issues/1483) to fix this. I will handle it in a subsequent PR.

ledwards2225 reviewed (approved):
Overall looks great. I'm slightly concerned about the quirkiness of the HidingIO (specific ordering of merge commitments / pairing points, reconstructing certain inputs twice). It's worth thinking if that can be avoided because its those subtle dependencies that made the old public input mechanism so tricky. Not a deal breaker though

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

ledwards2225 committed:
chore: remove public_inputs from DeciderVK (#15832)

Removes explicit storage of `public_inputs` in the DeciderVK. The public
inputs are folded as part of the wire polynomials, not as individual
elements so its not meaningful for the verifier accumulator to store
them explicitly (which has recently led to some confusion in IVC state
mocking). They are now stored directly in the oink/PG verifiers where
they are populated from the proof and used to reconstruct the IO data.

federicobarbacovi committed:
Added todo

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/add_merge_table_to_hiding_circuit_pub_inputs

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
chore: Update `UltraVerifier` api to return public inputs in the case of a MegaFlavor (#15868)

After
[#15382](https://github.com/AztecProtocol/aztec-packages/pull/15832),
Decider VKs do not contain public inputs anymore. In
[#15829](https://github.com/AztecProtocol/aztec-packages/pull/15829) we
need to access the public inputs to reconstruct the last merged table.
This PR moves the reconstruction inside the UltraVerifier verify
function and define two `verify` in two: one for `UltraFlavor` (which
returns `bool`) and one for `MegaFlavor` (which returns `bool` and the
last merged table).

Closes issue
[#1483](https://github.com/AztecProtocol/barretenberg/issues/1483).

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/add_merge_table_to_hiding_circuit_pub_inputs

federicobarbacovi committed:
Restructure HidingKernelIO

federicobarbacovi committed:
Update ultra verifier

federicobarbacovi committed:
Update CIVC

federicobarbacovi committed:
More CIVC

federicobarbacovi committed:
Fix databus tests

federicobarbacovi committed:
Fix mega tests

federicobarbacovi committed:
Fix protogalaxy tests

federicobarbacovi committed:
Fix ultra rec verifier tests except MegaZKFlavor

federicobarbacovi committed:
Clean up

federicobarbacovi committed:
Fix last test

federicobarbacovi committed:
Fix honk rec constraint tests

federicobarbacovi committed:
Fix Goblin AVM

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Fix Mega transcript tests

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Fix honk rec constraints

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/add_merge_table_to_hiding_circuit_pub_inputs

federicobarbacovi committed:
Squashed commit of the following:

commit 3dbca0382e25a0f1cf39f5699b16d521d3cd440c
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Thu Jul 24 13:24:16 2025 +0000

    Typos

commit 6a8a52300040159d60466298f87ef50e09179cf8
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Thu Jul 24 13:11:25 2025 +0000

    Fixes

commit b9e62d60c76ef96e521d6c89efd8266e1c13d414
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Thu Jul 24 12:15:59 2025 +0000

    Update remaining files

commit 37088cd2333c7dff59f3f45fca91903c5b093709
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Thu Jul 24 11:47:45 2025 +0000

    Update Goblin boomerang test

commit 39ac8c27bcf61a9bb8e3d947907386559750f4cc
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Thu Jul 24 11:46:02 2025 +0000

    Update goblin

commit 888fc823b3cd593a4c3fbf893c088682f74abf90
Author: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Date:   Thu Jul 24 11:26:40 2025 +0000

    Update merge verifier api

commit 4fb769b832204fe41aafc52745dcb4827db833db
Merge: eaf251ec5b f70e99d0ce
Author: AztecBot <tech@aztecprotocol.com>
Date:   Thu Jul 24 09:38:03 2025 +0000

    Merge branch 'next' into merge-train/barretenberg

commit f70e99d0ced51f09e1b30c43cfd866b17652f067
Author: Nicolás Venturo <nicolas.venturo@gmail.com>
Date:   Thu Jul 24 05:55:55 2025 -0300

    docs: updated code docs (#15701)

    With `Empty` being a supertrait of `Eq`, `Empty + Eq` becomes
    meaningless - this was emitting warnings.

    @LeilaWang what do you think about setting up a CI job that checks the
    protocol circuit crates emit no warnings? That's what we use in aztec-nr
    and it's been working wonderfully. tldr you need to run `nargo check
    --deny-warnings`.

federicobarbacovi committed:
Update CIVC

federicobarbacovi committed:
Introduce default method for ecc op tables

federicobarbacovi committed:
Reset Cargo lock

federicobarbacovi committed:
Fixes

federicobarbacovi committed:
Use IO mechanism for clearer handling of public inputs propagation

federicobarbacovi committed:
Fix

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/add_merge_table_to_hiding_circuit_pub_inputs

federicobarbacovi committed:
Minore chores

federicobarbacovi committed:
Fix

federicobarbacovi committed:
Fix

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/honk_recursion_constraint.cpp:
```diff
@@ -215,7 +216,7 @@ void create_dummy_vkey_and_proof(typename Flavor::CircuitBuilder& builder,
 HonkRecursionConstraintOutput<typename Flavor::CircuitBuilder> create_honk_recursion_constraints(
     typename Flavor::CircuitBuilder& builder, const RecursionConstraint& input, bool has_valid_witness_assignments)
-    requires IsRecursiveFlavor<Flavor>
+    requires(IsRecursiveFlavor<Flavor> && IsUltraHonk<typename Flavor::NativeFlavor>)
```
Added because this function is used only when `IsUltraHonk` is true. It avoids confusion now that the public inputs for `MegaFlavor` are different

federicobarbacovi committed:
Reset circuits

federicobarbacovi committed:
Default value of ecc ops reflect those obtained by finalizing an empty circuit

iakovenkos commented on barretenberg/cpp/src/barretenberg/stdlib/special_public_inputs/special_public_inputs.hpp:
```diff
@@ -126,9 +140,93 @@ template <typename Builder> class DefaultIO {
+        builder->finalize_public_inputs();
+    }
+
+    static std::array<G1, Builder::NUM_WIRES> empty_ecc_op_tables(Builder& builder)
```
can delete now, right?

federicobarbacovi replied:
We could, but in some places we will need to instantiate ecc op tables to the points at infinity as the starting point of the Goblin recursion. More precisely, we'll need to do it in the init Kernel, and in the Goblin AVM verifier. I thought it was nicer to have a method that produces such tables rather than having a loop every time

iakovenkos reviewed (approved):
Looks great, thanks for avoiding points at infinity in the default ecc tables!
